crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -180,48 +180,57 @@ impl Solution {
     /// Check that the sum of tokens entering the settlement is not less than
     /// the sum of tokens exiting the settlement.
     fn verify_asset_flow(&self) -> Result<(), VerificationError> {
-        let mut sum_inputs = eth::U256::zero();
-        let mut sum_outputs = eth::U256::zero();
-        for interaction in self.interactions.iter() {
-            match interaction {
-                Interaction::Custom(interaction) => {
-                    for input in interaction.inputs.iter() {
-                        sum_inputs += self
-                            .reference_amount(input)
-                            .ok_or(VerificationError::AssetFlow)?;
-                    }
-                    for output in interaction.outputs.iter() {
-                        sum_outputs += self
-                            .reference_amount(output)
-                            .ok_or(VerificationError::AssetFlow)?;
-                    }
-                }
-                Interaction::Liquidity(interaction) => {
-                    sum_inputs += self
-                        .reference_amount(&interaction.input)
-                        .ok_or(VerificationError::AssetFlow)?;
-                    sum_outputs += self
-                        .reference_amount(&interaction.output)
-                        .ok_or(VerificationError::AssetFlow)?;
-                }
-            }
-        }
+        // Sum settlement inputs across all interactions.
+        let inputs = self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.inputs().into_iter());
+        let sum_inputs = inputs
+            .fold(Some(eth::U256::zero()), |result, input| {
+                result.and_then(|sum| Some(sum + self.reference_amount(&input)?))
+            })
+            .ok_or(VerificationError::AssetFlow)?;
+
+        // Sum settlement outputs across all interactions.
+        let outputs = self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.outputs().into_iter());
+        let sum_outputs = outputs
+            .fold(Some(eth::U256::zero()), |result, output| {
+                result.and_then(|sum| Some(sum + self.reference_amount(&output)?))
+            })
+            .ok_or(VerificationError::AssetFlow)?;
+
+        // Check that the settlement doesn't steal tokens from the contract.
         if sum_inputs >= sum_outputs {
             Ok(())
         } else {
             Err(VerificationError::AssetFlow)
         }
     }
 
+    /// Check that internalized interactions only use trusted tokens.
```

Comment:
The idea of internalization is that the settlement contract poses as a liquidity pool to save on fees. This is only supposed to happen for realistic trades which means the solver has to bring proof that this price is achievable on-chain. Proof would be if we can successfully simulate the to-be-internalized transaction to validate that the trade would be legit.

Reply:
AFAIK currently only a single simulation is happening per solution. Not sure if it's the internalized or uninternalized version.

Reply:
The reasoning was already outlined [here](https://github.com/cowprotocol/services/issues/1009#issuecomment-1387621803). Internalization is only legal if the interaction could also be executed on-chain. If you internalize an interaction illegally you would get slashed.
To keep solvers from submitting illegally internalized solutions we should also simulate the uninternalized version so we know that every interaction can be executed on-chain and is therefore allowed to be internalized (if all the `input` tokens are trusted).

Reply:
This would be a sane default. I think if there is a reason solvers why solvers would want to submit a illegal solution knowingly (maybe it's actually more profitable that way for them 🤷‍♂️) it would be more appropriate to make them opt-out of that check with a run-time flag.